src/Goteo/Library/PostPrivacyAccess.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Goteo\Library;
+
+use Goteo\Application\Session;
+use Goteo\Model\Blog\Post;
+use Goteo\Model\Blog\Post\PostRewardAccess;
+use Goteo\Model\Invest;
+use Goteo\Model\User;
+
+class PostPrivacyAccess
+{
+    static public function canAccess(?User $user, Post $post): bool
+    {
+        $admin = Session::isAdmin();
+        if ($admin)
+            return true;
+
+        if ($user instanceof User) {
+            return self::canUserAccess($user, $post);
+        }
+
+        if ($post->access_limited)
+            return false;
+
+        if ($post->publish) return true;
+
+        return false;
+    }
+
+    private function canUserAccess(User $user, Post $post): bool
+    {
+        if ($user->id == $post->author)
+            return true;
+
+        if ($post->access_limited) {
+            $postRewardAccessList = PostRewardAccess::getList(['post_id' => $post->id]);
+            foreach($postRewardAccessList as $postRewardAccess) {
+                $reward = $postRewardAccess->getReward();
+                if ($reward->subscribable) {
+                    $hasAccessToPost = Invest::hasUserInvestToRewardInTheLastMonth($user, $reward);
+                    if ($hasAccessToPost)
```

Comment:
Deciding to return `true` only when `$hasAccessToPost` is `true` makes the function continue to the next `else` statement, making it possible at this part of the code to obtain an erroneous `true` value when trying to determine if a user has access to a subscription-based reward when it skips to the method `hasUserInvestToReward` which returns `true` on any amount of successful invests regardless of the time window.

I think this functionality can be accomplished much more elegantly by:
1. Refactor `hasUserInvestToReward` to hold the logic to check if the given `$reward` is a subscription-based one, then do the necessary lookup for that case inside the method.
2. Simply return the value of the method `hasUserInvestToReward`, e.g:
```php
$reward = $postRewardAccess->getReward();
return Invest::hasUserInvestToReward($reward);
```

Reply:
I don't get the part about the falsy true, I don't get how it will jump to the next `else`, since the `if` is about the subcribability of the reward :S

Apart from that, i completely agree about the refactoring!

Reply:
My bad, as I was reading the if I thought it would skip due to a syntax misreading. Please do refactor it to avoid `if ($hasAccessToPost)` and simply return the value of the privacy check method.

Regarding on where to put the logic, I'm OK with the two methods we discussed, for reference:

1. Create an InvestRepository to hold the SQL to search for invests to specific rewards.
2. Use an approach similar to Symfony's Voters: https://symfony.com/doc/current/security/voters.html